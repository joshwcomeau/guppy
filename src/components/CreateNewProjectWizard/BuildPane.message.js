// @flow
import { defineMessages } from 'react-intl';

const messages = {
  step: {
    installingCliTool: {
      id: 'createNewProjectWizard.buildPane.step.installingCliTool',
      defaultMessage: 'Installing build tool',
      description:
        'Text "Installing build tool" for the installingCliTool step',
    },
    creatingProjectDirectory: {
      id: 'createNewProjectWizard.buildPane.step.creatingProjectDirectory',
      defaultMessage: 'Creating project directory',
      description:
        'Text "Creating project directory" for the creatingProjectDirectory step',
    },
    installingDependencies: {
      id: 'createNewProjectWizard.buildPane.step.installingDependencies',
      defaultMessage: 'Installing dependencies',
      description:
        'Text "Installing dependencies" for the installingDependencies step',
    },
    guppification: {
      id: 'createNewProjectWizard.buildPane.step.guppification',
      defaultMessage: 'Persisting project info',
      description: 'Text "Persisting project info" for the guppification step',
    },
    takeAWhile: {
      id: 'createNewProjectWizard.buildPane.step.takeAWhile',
      defaultMessage: 'This step can take a while...',
      description: 'Text to tell the user the step might take a while',
    },
  },
  main: {
    created: {
      id: 'createNewProjectWizard.buildPane.main.created',
      defaultMessage: 'Project Created!',
      description: 'Text when the project is created',
    },
    building: {
      id: 'createNewProjectWizard.buildPane.main.building',
      defaultMessage: 'Building Project...',
      description: 'Text when the project is being built',
    },
  },
  error: {
    insufficientData: {
      id: 'createNewProjectWizard.buildPane.error.insufficientData',
      defaultMessage:
        'Tried to build project with insufficient data. See console for more info',
      description: 'Error message when data is insufficient',
    },
  },
};

const definedMessages = defineMessages(messages);

export default definedMessages;
