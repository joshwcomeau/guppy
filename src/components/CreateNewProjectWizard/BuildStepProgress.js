// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import { check } from 'react-icons-kit/feather/check';

import Spinner from '../Spinner';
import { COLORS } from '../../constants';
import FadeIn from '../FadeIn';

type Status = 'upcoming' | 'in-progress' | 'done';

type Props = {
  step: any,
  status: Status,
};

type State = {
  shouldShowAdditionalCopy: boolean,
};

class BuildStepProgress extends PureComponent<Props, State> {
  timeoutId: number;

  state = {
    shouldShowAdditionalCopy: false,
  };

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.status === 'done') {
      this.setState({ shouldShowAdditionalCopy: false });
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.status !== 'in-progress' &&
      this.props.status === 'in-progress'
    ) {
      this.timeoutId = window.setTimeout(this.triggerAdditionalCopy, 5000);
    }
  }

  componentWillUnmount() {
    window.clearTimeout(this.timeoutId);
  }

  triggerAdditionalCopy = () => {
    this.setState({ shouldShowAdditionalCopy: true });
  };

  render() {
    const { step, status } = this.props;
    const { shouldShowAdditionalCopy } = this.state;
    return (
      <Wrapper isUpcoming={status === 'upcoming'}>
        <IconWrapper>
          {status === 'in-progress' && <Spinner size={20} />}
          {status === 'done' && <Checkmark size={32} icon={check} />}
        </IconWrapper>
        <ChildWrapper>
          <MainCopy>{step.copy}</MainCopy>
          {step.additionalCopy &&
            shouldShowAdditionalCopy && (
              <FadeIn>
                <AdditionalCopy>{step.additionalCopy}</AdditionalCopy>
              </FadeIn>
            )}
        </ChildWrapper>
      </Wrapper>
    );
  }
}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 10px;
  font-size: 18px;
  opacity: ${props => (props.isUpcoming ? 0.5 : 1)};
  will-change: opacity;
  transition: opacity 800ms;
`;

const IconWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 32px;
  height: 32px;
  margin-right: 10px;
`;

const ChildWrapper = styled.div`
  position: relative;
  flex: 1;
`;

const Checkmark = styled(IconBase)`
  color: ${COLORS.lightSuccess};
`;

const MainCopy = styled.div`
  font-weight: 600;
  -webkit-font-smoothing: antialiased;
  font-size: 18px;
`;

const AdditionalCopy = styled.div`
  position: absolute;
  top: 110%;
  font-size: 14px;
`;

export default BuildStepProgress;
