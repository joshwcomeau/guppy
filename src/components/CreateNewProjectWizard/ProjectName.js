// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { Tooltip } from 'react-tippy';
import IconBase from 'react-icons-kit';
import { u2728 as sparkles } from 'react-icons-kit/noto_emoji_regular/u2728';

import { COLORS } from '../../constants';
import { generateRandomName } from '../../services/project-name.service';
import { range, random } from '../../utils';

import FormField from '../FormField';
import TextInput from '../TextInput';
import HoverableOutlineButton from '../HoverableOutlineButton';

type Props = {
  name: string,
  isFocused: boolean,
  isProjectNameTaken: boolean,
  // TODO: Change to onFocus, onBlur, onChange for consistency
  handleFocus: () => void,
  handleBlur: () => void,
  handleChange: (val: string) => void,
  handleSubmit: () => Promise<any> | void,
};

type State = {
  randomizedOverrideName: ?string,
  randomizeCount: number,
  showRandomizeTooltip: boolean,
};

class ProjectName extends PureComponent<Props, State> {
  state = {
    randomizedOverrideName: null,
    randomizeCount: 0,
    showRandomizeTooltip: false,
  };

  node: HTMLElement;
  timeoutId: number;

  componentDidMount() {
    this.node.focus();

    this.timeoutId = window.setTimeout(() => {
      this.setState({ showRandomizeTooltip: true });
    }, 3500);
  }

  componentWillUnmount() {
    window.clearTimeout(this.timeoutId);
  }

  getRidOfRandomizeTooltip = () => {
    // If it hasn't been shown yet, ensure it won't be.
    window.clearTimeout(this.timeoutId);

    if (this.state.showRandomizeTooltip) {
      this.setState({ showRandomizeTooltip: false });
    }
  };

  handleRandomize = () => {
    const { randomizeCount } = this.state;
    const newName = generateRandomName();

    this.getRidOfRandomizeTooltip();

    this.props.handleChange(newName);

    // See note above. $FlowFixMe
    this.node.focus();

    let numOfTicks;
    if (randomizeCount <= 3) {
      numOfTicks = 10;
    } else if (randomizeCount <= 6) {
      numOfTicks = 3;
    } else {
      numOfTicks = 0;
    }

    if (numOfTicks > 0) {
      this.scramble(this.props.name, newName, numOfTicks);
    }
  };

  updateName = (ev: SyntheticInputEvent<*>) => {
    this.getRidOfRandomizeTooltip();

    this.props.handleChange(ev.target.value);
  };

  scramble = (from: string, to: string, numOfTicks: number) => {
    const toNumOfChars = to.length;

    // Start with all spaces revealed
    let revealedLetterIndices = to
      .split('')
      .map((_, i) => i)
      .filter(i => to[i].match(/\s/));

    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';

    const tick = (i = 0) => {
      const progress = i / numOfTicks;

      const isFinished =
        this.state.randomizedOverrideName === to || progress === 1;

      if (isFinished) {
        this.setState({
          randomizedOverrideName: null,
          randomizeCount: this.state.randomizeCount + 1,
        });
        return;
      }

      // As we get closer to the "to" term, letters should be progressively
      // revealed. So we may not want to randomize every letter.
      const numOfRevealedLetters = revealedLetterIndices.length;
      const numOfLettersToReveal = Math.floor(progress * toNumOfChars);

      range(0, numOfLettersToReveal - numOfRevealedLetters).forEach(() => {
        // We want to pick a random index to reveal
        // TODO: Prioritize early-in-the-string letters, for a nicer effect?
        // TODO: Worry about duplicates, so that the timing is more consistent?
        revealedLetterIndices.push(random(0, toNumOfChars));
      });

      const semiRandomString = to
        .split('')
        .map((letter, index) => {
          if (revealedLetterIndices.includes(index)) {
            return letter;
          }

          return letters.charAt(random(0, letters.length));
        })
        .join('');

      this.setState({ randomizedOverrideName: semiRandomString });

      window.setTimeout(() => tick(i + 1), 15 + progress * 100);
    };

    tick();
  };

  maybeHandleSubmit = (ev: SyntheticKeyboardEvent<*>) => {
    if (ev.key === 'Enter') {
      this.props.handleSubmit();

      // See note above. $FlowFixMe
      this.node.blur();
    }
  };

  setRef = (node: HTMLElement) => (this.node = node);

  render() {
    const {
      name,
      isFocused,
      isProjectNameTaken,
      handleFocus,
      handleBlur,
    } = this.props;
    const { randomizedOverrideName, showRandomizeTooltip } = this.state;

    return (
      <FormField
        useLabelTag
        label="Project Name"
        isFocused={isFocused}
        hasError={isProjectNameTaken}
        spacing={15}
      >
        <TextInput
          innerRef={this.setRef}
          type="text"
          value={randomizedOverrideName || name}
          isFocused={isFocused}
          hasError={isProjectNameTaken}
          onChange={this.updateName}
          onFocus={handleFocus}
          onBlur={handleBlur}
          onKeyPress={this.maybeHandleSubmit}
          placeholder="Your Amazing Project Name"
        >
          <ButtonPositionAdjuster>
            <Tooltip
              arrow
              open={showRandomizeTooltip}
              title="Generate a random code-name"
              delay={100}
              distance={24}
              duration={200}
              animation="shift"
              trigger="manual"
            >
              <HoverableOutlineButton
                noPadding
                onMouseDown={() => window.requestAnimationFrame(handleFocus)}
                onClick={this.handleRandomize}
                style={{ width: 32, height: 32 }}
              >
                <IconBase size={22} icon={sparkles} />
              </HoverableOutlineButton>
            </Tooltip>
          </ButtonPositionAdjuster>
        </TextInput>
        {isProjectNameTaken && (
          <ErrorMessage>
            Sorry, a project with this name already exists.
          </ErrorMessage>
        )}
      </FormField>
    );
  }
}

export const ErrorMessage = styled.div`
  margin-top: 6px;
  color: ${COLORS.error};
`;

const ButtonPositionAdjuster = styled.div`
  transform: translateY(5px);
`;

export default ProjectName;
