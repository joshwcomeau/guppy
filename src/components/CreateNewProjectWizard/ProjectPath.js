// @flow
import path from 'path';
import { remote } from 'electron';
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { Tooltip } from 'react-tippy';
import styled from 'styled-components';

import { changeDefaultProjectPath } from '../../actions';
import { getDefaultProjectPath } from '../../reducers/app-settings.reducer';
import { getProjectNameSlug } from '../../services/create-project.service';
import { COLORS } from '../../constants';

import TextButton from '../TextButton';

type Props = {
  projectHome: string,
  projectName: string,
  changeDefaultProjectPath: (path: string) => void,
};

class ProjectPath extends PureComponent<Props> {
  updatePath = () => {
    remote.dialog.showOpenDialog(
      {
        message: 'Select the directory of Project',
        properties: ['openDirectory'],
      },
      (paths: ?[string]) => {
        // The user might cancel out without selecting a directory.
        // In that case, do nothing.
        if (!paths) {
          return;
        }

        // Only a single path should be selected
        const [firstPath] = paths;
        this.props.changeDefaultProjectPath(firstPath);
      }
    );
  };

  render() {
    const { projectHome, projectName } = this.props;

    const projectNameSlug = getProjectNameSlug(projectName);

    // Join the projectHome with the prospective project ID
    // Hide the leading forward-slash, on Mac/Linux
    const fullProjectPath = path
      .join(projectHome, projectNameSlug)
      .replace(/^\//, '');

    // Using CSS text-overflow is proving challenging, so we'll just crop it
    // with JS.
    const CLAMP_AT = 29;
    let displayedProjectPath = fullProjectPath;
    if (displayedProjectPath.length > CLAMP_AT) {
      displayedProjectPath = `${displayedProjectPath.slice(0, CLAMP_AT - 1)}â€¦`;
    }

    return (
      <MainText>
        Project will be created in{' '}
        <Tooltip title={fullProjectPath} position="bottom">
          <DirectoryButton onClick={() => this.updatePath()}>
            {displayedProjectPath}
          </DirectoryButton>
        </Tooltip>
      </MainText>
    );
  }
}

const MainText = styled.div`
  text-align: left;
  margin: -20px 0 30px 5px;
  font-size: 15px;
  color: ${COLORS.gray[400]};
`;

const DirectoryButton = styled(TextButton)`
  font-family: 'Fira Mono';
  font-size: 12px;
  color: ${COLORS.gray[600]};
  text-decoration: none;
`;

const mapStateToProps = state => ({
  projectHome: getDefaultProjectPath(state),
});

const mapDispatchToProps = {
  changeDefaultProjectPath,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProjectPath);
