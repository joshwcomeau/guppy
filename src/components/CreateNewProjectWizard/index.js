// @flow
export { default } from './CreateNewProjectWizard';
