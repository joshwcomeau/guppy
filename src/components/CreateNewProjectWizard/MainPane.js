// @flow
import React, { PureComponent, Fragment } from 'react';
import { Motion, spring } from 'react-motion';
import styled from 'styled-components';

import reactIconSrc from '../../assets/images/react-icon.svg';
import gatsbyIconSrc from '../../assets/images/gatsby_small.png';
import nextjsIconSrc from '../../assets/images/nextjs_small.png';

import FormField from '../FormField';
import ProjectIconSelection from '../ProjectIconSelection';
import FadeIn from '../FadeIn';

import ProjectName from './ProjectName';
import ProjectPath from './ProjectPath';
import SubmitButton from './SubmitButton';
import ButtonWithIcon from '../ButtonWithIcon';
import Spacer from '../Spacer';

import type { Field, Status } from './types';
import type { ProjectType } from '../../types';

type Props = {
  projectName: string,
  projectType: ?ProjectType,
  projectIcon: ?string,
  activeField: ?Field,
  status: Status,
  currentStepIndex: number,
  hasBeenSubmitted: boolean,
  isProjectNameTaken: boolean,
  updateFieldValue: (field: Field, value: any) => void,
  focusField: (field: ?Field) => void,
  handleSubmit: () => Promise<any> | void,
};

class MainPane extends PureComponent<Props> {
  handleFocusProjectName = () => this.props.focusField('projectName');
  handleBlurProjectName = () => this.props.focusField(null);

  updateProjectName = (projectName: string) =>
    this.props.updateFieldValue('projectName', projectName);
  updateProjectType = (projectType: ProjectType) =>
    this.props.updateFieldValue('projectType', projectType);
  updateProjectIcon = (projectIcon: string) =>
    this.props.updateFieldValue('projectIcon', projectIcon);

  render() {
    const {
      projectName,
      projectType,
      projectIcon,
      activeField,
      currentStepIndex,
      hasBeenSubmitted,
      isProjectNameTaken,
      handleSubmit,
    } = this.props;

    return (
      <Fragment>
        <Motion style={{ offset: spring(currentStepIndex === 0 ? 50 : 0) }}>
          {({ offset }) => (
            <Wrapper style={{ transform: `translateY(${offset}px)` }}>
              <ProjectName
                name={projectName}
                isFocused={activeField === 'projectName'}
                handleFocus={this.handleFocusProjectName}
                handleBlur={this.handleBlurProjectName}
                handleChange={this.updateProjectName}
                handleSubmit={handleSubmit}
                isProjectNameTaken={isProjectNameTaken}
              />
              <ProjectPath projectName={projectName} />

              {currentStepIndex > 0 && (
                <FadeIn>
                  <FormField
                    label="Project Type"
                    isFocused={activeField === 'projectType'}
                  >
                    <ProjectTypeTogglesWrapper>
                      {/* Todo: Make it easier to add new flows - e.g. map over an array to generate the UI*/}
                      <ButtonWithIcon
                        showStroke={projectType === 'create-react-app'}
                        icon={<ReactIcon src={reactIconSrc} />}
                        onClick={() =>
                          this.updateProjectType('create-react-app')
                        }
                      >
                        Vanilla React
                      </ButtonWithIcon>
                      <Spacer inline size={10} />
                      <ButtonWithIcon
                        showStroke={projectType === 'gatsby'}
                        icon={<GatsbyIcon src={gatsbyIconSrc} />}
                        onClick={() => this.updateProjectType('gatsby')}
                      >
                        Gatsby
                      </ButtonWithIcon>
                      <Spacer inline size={10} />
                      <ButtonWithIcon
                        showStroke={projectType === 'nextjs'}
                        icon={<NextjsIcon src={nextjsIconSrc} />}
                        onClick={() => this.updateProjectType('nextjs')}
                      >
                        Next.js
                      </ButtonWithIcon>
                    </ProjectTypeTogglesWrapper>
                  </FormField>
                </FadeIn>
              )}

              {currentStepIndex > 1 && (
                <FadeIn>
                  <FormField
                    label="Project Icon"
                    focusOnClick={false}
                    isFocused={activeField === 'projectIcon'}
                  >
                    <ProjectIconSelection
                      selectedIcon={projectIcon}
                      randomize={true}
                      limitTo={8}
                      onSelectIcon={this.updateProjectIcon}
                    />
                  </FormField>
                </FadeIn>
              )}
            </Wrapper>
          )}
        </Motion>
        <SubmitButtonWrapper>
          <SubmitButton
            isDisabled={
              isProjectNameTaken ||
              !projectName ||
              (currentStepIndex > 0 && !projectType) ||
              (currentStepIndex > 1 && !projectIcon)
            }
            readyToBeSubmitted={currentStepIndex >= 2}
            hasBeenSubmitted={hasBeenSubmitted}
            onSubmit={handleSubmit}
          />
        </SubmitButtonWrapper>
      </Fragment>
    );
  }
}

const Wrapper = styled.div`
  height: 500px;
  will-change: transform;
`;

const ReactIcon = styled.img`
  width: 32px;
  height: 32px;
`;

const GatsbyIcon = styled.img`
  width: 22px;
  height: 22px;
`;

const NextjsIcon = styled.img`
  width: 22px;
  height: 22px;
`;

const ProjectTypeTogglesWrapper = styled.div`
  margin-top: 8px;
  margin-left: -8px;
`;

const SubmitButtonWrapper = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 30px;
  text-align: center;
`;

export default MainPane;
