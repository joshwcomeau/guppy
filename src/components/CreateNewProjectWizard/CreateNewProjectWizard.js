// @flow
import React, { Fragment, PureComponent } from 'react';
import { connect } from 'react-redux';
import Transition from 'react-transition-group/Transition';
import { remote } from 'electron';

import * as actions from '../../actions';
import {
  getAppSettings,
  getDefaultProjectPath,
} from '../../reducers/app-settings.reducer';
import { getById } from '../../reducers/projects.reducer';
import { getOnlineState } from '../../reducers/app-status.reducer';
import { getOnboardingCompleted } from '../../reducers/onboarding-status.reducer';
import { getProjectNameSlug } from '../../services/create-project.service';
import { checkIfProjectExists } from '../../services/create-project.service';
import { urlExists } from '../../services/check-if-url-exists.service';
import { replaceProjectStarterStringWithUrl } from './helpers';

import TwoPaneModal from '../TwoPaneModal';
import Debounced from '../Debounced';

import MainPane from './MainPane';
import SummaryPane from './SummaryPane';
import BuildPane from './BuildPane';

import type { Field, Status, Step } from './types';

import type { ProjectType, ProjectInternal, AppSettings } from '../../types';
import type { Dispatch } from '../../actions/types';

export const FORM_STEPS: Array<Field> = [
  'projectName',
  'projectType',
  'projectIcon',
  'projectStarter',
];

export const dialogOptionsFolderExists = {
  type: 'warning',
  title: 'Project directory exists',
  message:
    "Looks like there's already a project with that name. Did you mean to import it instead?",
  buttons: ['OK'],
};

export const dialogCallbackFolderExists = (
  resolve: (result: any) => void,
  reject: (error: any) => void
) => (result: number) => {
  if (result === 0) {
    return reject();
  }

  resolve();
};

export const dialogStarterNotFoundErrorArgs = (projectStarter: string) => [
  `Starter ${projectStarter} not found`,
  'Please check your starter url or use the starter selection to pick a starter.',
];

const { dialog } = remote;

type Props = {
  settings: AppSettings,
  projects: {
    [projectId: string]: ProjectInternal,
  },
  projectHomePath: string,
  isVisible: boolean,
  isOnboardingCompleted: boolean,
  addProject: Dispatch<typeof actions.addProject>,
  createNewProjectCancel: Dispatch<typeof actions.createNewProjectCancel>,
  createNewProjectFinish: Dispatch<typeof actions.createNewProjectFinish>,
  isOnline: boolean,
};

type State = {
  projectName: string,
  projectType: ?ProjectType,
  projectIcon: ?string,
  projectStarter: string,
  activeField: ?Field,
  settings: ?AppSettings,
  status: Status,
  currentStep: Step,
  isProjectNameTaken: boolean,
};

const initialState = {
  projectName: '',
  projectType: null,
  projectIcon: null,
  projectStarter: '',
  activeField: 'projectName',
  status: 'filling-in-form',
  currentStep: 'projectName',
  isProjectNameTaken: false,
  settings: null,
};

export class CreateNewProjectWizard extends PureComponent<Props, State> {
  state = initialState;
  timeoutId: number;

  componentDidMount() {
    this.reinitialize(); // needed to load app settings
  }

  componentWillUnmount() {
    window.clearTimeout(this.timeoutId);
  }

  updateFieldValue = (field: Field, value: any) => {
    this.setState({
      [field]: value,
      activeField: field,
    });

    if (field === 'projectName') {
      this.verifyProjectNameUniqueness(value);
    }
  };

  focusField = (field: ?Field) => {
    this.setState({
      activeField: field,
    });
  };

  verifyProjectNameUniqueness = (name: string) => {
    const { projects } = this.props;

    // Check to see if this name is already taken
    const slug = getProjectNameSlug(name);
    const isAlreadyTaken = !!Object.keys(this.props.projects).some(
      id => projects[id].name === slug
    );

    if (isAlreadyTaken) {
      this.setState({
        isProjectNameTaken: true,
      });
      return;
    }

    // If this update fixes the problem, unset the error status
    if (!isAlreadyTaken && this.state.isProjectNameTaken) {
      this.setState({
        isProjectNameTaken: false,
      });
    }
  };

  checkProjectLocationUsage = () => {
    return new Promise<any>((resolve, reject) => {
      const projectName = getProjectNameSlug(this.state.projectName);
      if (checkIfProjectExists(this.props.projectHomePath, projectName)) {
        // show warning that the project folder already exists & stop creation
        dialog.showMessageBox(
          dialogOptionsFolderExists,
          dialogCallbackFolderExists(resolve, reject)
        );
      } else {
        resolve();
      }
    });
  };

  checkIfStarterUrlExists = async () => {
    const { projectStarter: projectStarterInput } = this.state;
    // Add url to starter if not passed & not an empty string
    const projectStarter = replaceProjectStarterStringWithUrl(
      projectStarterInput
    );

    if (projectStarter === '') {
      return;
    }

    const exists = await urlExists(projectStarter);

    if (!exists) {
      // starter not found
      dialog.showErrorBox(...dialogStarterNotFoundErrorArgs(projectStarter));
      throw new Error('starter-not-found');
    }
  };

  handleSubmit = () => {
    const currentStepIndex = FORM_STEPS.indexOf(this.state.currentStep);
    const nextStep = FORM_STEPS[currentStepIndex + 1];
    const lastStep = this.state.projectType === 'gatsby' ? 3 : 2;

    if (currentStepIndex >= lastStep) {
      return Promise.all([
        this.checkProjectLocationUsage(),
        this.checkIfStarterUrlExists(),
      ])
        .then(() => {
          // not in use & starter exists (if Gatsby project)
          this.setState({
            activeField: null,
            status: 'building-project',
          });
        })
        .catch(() => {
          // Swallow any errors, as the promise above will have shown the appropriate dialog on error
        });
    }

    this.setState({
      currentStep: nextStep,
      activeField: nextStep,
    });
  };

  finishBuilding = (project: ProjectInternal) => {
    const { isOnboardingCompleted, projectHomePath } = this.props;
    const { projectType } = this.state;

    // Should be impossible
    if (!projectType) {
      throw new Error('Project created without projectType');
    }

    this.props.createNewProjectFinish();

    this.timeoutId = window.setTimeout(() => {
      this.props.addProject(
        project,
        projectHomePath,
        projectType,
        isOnboardingCompleted
      );

      this.timeoutId = window.setTimeout(this.reinitialize, 500);
    }, 500);
  };

  reinitialize = () => {
    const { settings } = this.props;

    this.setState({
      ...initialState,
      projectType: settings.general.defaultProjectType,
    });
  };

  render() {
    const {
      isVisible,
      createNewProjectCancel,
      projectHomePath,
      isOnline,
    } = this.props;
    const {
      projectName,
      projectType,
      projectIcon,
      projectStarter,
      activeField,
      status,
      currentStep,
      isProjectNameTaken,
    } = this.state;

    const project = {
      projectName,
      projectType,
      projectIcon,
      projectStarter,
    };

    const readyToBeBuilt = status !== 'filling-in-form';

    return (
      <Transition in={isVisible} timeout={300}>
        {transitionState => (
          <Fragment>
            <TwoPaneModal
              isFolded={readyToBeBuilt}
              transitionState={transitionState}
              isDismissable={status !== 'building-project'}
              onDismiss={createNewProjectCancel}
              leftPane={
                <Debounced on={activeField} duration={40}>
                  <SummaryPane
                    currentStep={currentStep}
                    activeField={activeField}
                    projectType={projectType}
                  />
                </Debounced>
              }
              rightPane={
                <MainPane
                  {...project}
                  status={status}
                  activeField={activeField}
                  currentStepIndex={FORM_STEPS.indexOf(currentStep)}
                  updateFieldValue={this.updateFieldValue}
                  focusField={this.focusField}
                  handleSubmit={this.handleSubmit}
                  hasBeenSubmitted={status !== 'filling-in-form'}
                  isProjectNameTaken={isProjectNameTaken}
                  isOnline={isOnline}
                />
              }
              backface={
                <BuildPane
                  {...project}
                  status={status}
                  projectHomePath={projectHomePath}
                  handleCompleteBuild={this.finishBuilding}
                />
              }
            />
          </Fragment>
        )}
      </Transition>
    );
  }
}

const mapStateToProps = state => ({
  projects: getById(state),
  projectHomePath: getDefaultProjectPath(state),
  isVisible: state.modal === 'new-project-wizard',
  isOnboardingCompleted: getOnboardingCompleted(state),
  settings: getAppSettings(state),
  isOnline: getOnlineState(state),
});

const mapDispatchToProps = {
  addProject: actions.addProject,
  createNewProjectCancel: actions.createNewProjectCancel,
  createNewProjectFinish: actions.createNewProjectFinish,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateNewProjectWizard);
