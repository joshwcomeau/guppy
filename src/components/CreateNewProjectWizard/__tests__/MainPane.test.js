import React from 'react';
import { shallow } from 'enzyme';
import { Spring } from 'react-spring';
import lolex from 'lolex';

import MainPane from '../MainPane';
import ProjectName from '../ProjectName';
import ProjectPath from '../ProjectPath';
import { FORM_STEPS } from '../CreateNewProjectWizard';

describe('MainPane component', () => {
  let wrapper;
  let instance;
  let mockActions;

  const project = {
    projectName: 'Project name',
    projectType: 'gatsby',
    projectIcon: 'an-icon',
    projectStarter: null,
  };

  const shallowRenderMainPane = (
    step,
    isProjectNameTaken = false,
    status = 'filling-in-form'
  ) => {
    mockActions = {
      updateFieldValue: jest.fn(),
      focusField: jest.fn(),
      handleSubmit: jest.fn(),
    };
    return shallow(
      <MainPane
        {...project}
        status={status}
        activeField={FORM_STEPS[step]}
        currentStepIndex={step}
        {...mockActions}
        hasBeenSubmitted={status !== 'filling-in-form'}
        isProjectNameTaken={isProjectNameTaken}
      />
    );
  };

  lolex.install();

  describe('Rendering', () => {
    const checkConditionalSteps = i => {
      wrapper = shallowRenderMainPane(i);
      instance = wrapper.instance();
      expect(instance.renderConditionalSteps(i)).toMatchSnapshot();
    };

    it('should render ProjectName & ProjectPath', () => {
      wrapper = shallowRenderMainPane(0)
        .find(Spring)
        .renderProp('children')({ offset: 50 });
      expect(wrapper.find(ProjectName).exists()).toBe(true);
      expect(wrapper.find(ProjectPath).exists()).toBe(true);
    });

    for (let i = 0; i < FORM_STEPS.length; i++) {
      it(`should render step ${FORM_STEPS[i]}`, () => checkConditionalSteps(i));
    }
  });
  describe('Component logic', () => {
    beforeEach(() => {
      wrapper = shallowRenderMainPane(0);
      instance = wrapper.instance();
    });

    it('should trigger projectName focus', () => {
      instance.handleFocusProjectName();
      expect(mockActions.focusField).toHaveBeenCalledWith('projectName');
    });

    it('should trigger projectName blur', () => {
      instance.handleBlurProjectName();
      expect(mockActions.focusField).toHaveBeenCalledWith(null);
    });

    it('should trigger projectStarter focus', () => {
      instance.handleFocusStarter();
      expect(mockActions.focusField).toHaveBeenCalledWith('projectStarter');
    });

    it('should update projectName', () => {
      instance.updateProjectName('New name');
      expect(mockActions.updateFieldValue).toHaveBeenCalledWith(
        'projectName',
        'New name'
      );
    });

    it('should update projectType', () => {
      instance.updateProjectType('create-react-app');
      expect(mockActions.updateFieldValue).toHaveBeenCalledWith(
        'projectType',
        'create-react-app'
      );
    });

    it('should update projectIcon', () => {
      instance.updateProjectIcon('new-icon');
      expect(mockActions.updateFieldValue).toHaveBeenCalledWith(
        'projectIcon',
        'new-icon'
      );
    });

    it('should update projectStarter', () => {
      instance.updateGatsbyStarter('gatsby-blog-starter');
      expect(mockActions.updateFieldValue).toHaveBeenCalledWith(
        'projectStarter',
        'gatsby-blog-starter'
      );
    });

    it('should disable submit if form not ready', () => {
      // Clear props
      wrapper.setProps({
        projectType: null,
        projectIcon: null,
      });

      let disabled = instance.isSubmitDisabled(3, 3);
      expect(disabled).toBeTruthy();
    });
  });
});
