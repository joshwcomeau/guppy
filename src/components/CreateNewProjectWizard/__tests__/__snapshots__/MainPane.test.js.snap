// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPane component Rendering should render step projectIcon 1`] = `
Object {
  "lastIndex": 3,
  "steps": Array [
    <styled.div>
      <FormField
        isFocused={false}
        label="Project Type"
        spacing={10}
      >
        <ProjectTypeSelection
          onProjectTypeSelect={[Function]}
          projectType="gatsby"
        />
      </FormField>
    </styled.div>,
    <styled.div>
      <FormField
        focusOnClick={false}
        isFocused={true}
        label="Project Icon"
        spacing={10}
      >
        <ProjectIconSelection
          limitTo={9}
          onSelectIcon={[Function]}
          randomize={true}
          selectedIcon="an-icon"
        />
      </FormField>
    </styled.div>,
  ],
}
`;

exports[`MainPane component Rendering should render step projectName 1`] = `
Object {
  "lastIndex": 3,
  "steps": Array [],
}
`;

exports[`MainPane component Rendering should render step projectStarter 1`] = `
Object {
  "lastIndex": 3,
  "steps": Array [
    <styled.div>
      <FormField
        isFocused={false}
        label="Project Type"
        spacing={10}
      >
        <ProjectTypeSelection
          onProjectTypeSelect={[Function]}
          projectType="gatsby"
        />
      </FormField>
    </styled.div>,
    <styled.div>
      <FormField
        focusOnClick={false}
        isFocused={false}
        label="Project Icon"
        spacing={10}
      >
        <ProjectIconSelection
          limitTo={9}
          onSelectIcon={[Function]}
          randomize={true}
          selectedIcon="an-icon"
        />
      </FormField>
    </styled.div>,
    <styled.div>
      <FormField
        isFocused={true}
        label="Project Starter"
        spacing={15}
      >
        <ProjectStarter
          handleFocus={[Function]}
          isFocused={true}
          onSelect={[Function]}
          projectStarter={null}
        />
      </FormField>
    </styled.div>,
  ],
}
`;

exports[`MainPane component Rendering should render step projectType 1`] = `
Object {
  "lastIndex": 3,
  "steps": Array [
    <styled.div>
      <FormField
        isFocused={true}
        label="Project Type"
        spacing={10}
      >
        <ProjectTypeSelection
          onProjectTypeSelect={[Function]}
          projectType="gatsby"
        />
      </FormField>
    </styled.div>,
  ],
}
`;
